src/agent.rs

```diff
@@ -43,79 +43,99 @@ impl Decoder for MessageCodec {
+        loop {
+            if let Some(incoming_message) = self.adapter.try_next().await? {
+                let response = self.agent.handle(incoming_message).await.map_err(|e| {
+                    error!("Error handling message; error = {:?}", e);
+                    AgentError::User
+                })?;
+
+                self.adapter.send(response).await?;
+            } else {
+                // Reached EOF of the stream (client disconnected),
+                // we can close the socket and exit the handler.
+                return Ok(());
+            }
+        }
     }
+}
 
-    fn run_unix(self, path: impl AsRef<Path>) -> Result<(), Box<dyn Error + Send + Sync>> {
-        self.run_listener(UnixListener::bind(path)?)
+#[async_trait]
+pub trait ListeningSocket {
+    type Stream: fmt::Debug + AsyncRead + AsyncWrite + Send + Unpin + 'static;
+
+    async fn accept(&self) -> io::Result<Self::Stream>;
+}
+
+#[async_trait]
+impl ListeningSocket for UnixListener {
+    type Stream = UnixStream;
+    async fn accept(&self) -> io::Result<Self::Stream> {
+        UnixListener::accept(self).await.map(|(s, _addr)| s)
     }
+}
 
-    #[allow(clippy::unit_arg)]
-    fn run_tcp(self, addr: &str) -> Result<(), Box<dyn Error + Send + Sync>> {
-        let socket = TcpListener::bind(&addr.parse::<SocketAddr>()?)?;
-        Ok(tokio::run(handle_clients!(self, socket)))
+#[async_trait]
+impl ListeningSocket for TcpListener {
+    type Stream = TcpStream;
+    async fn accept(&self) -> io::Result<Self::Stream> {
+        TcpListener::accept(self).await.map(|(s, _addr)| s)
     }
+}
 
-    #[allow(clippy::unit_arg)]
-    fn listen(self, socket: service_binding::Listener) -> Result<(), Box<dyn Error + Send + Sync>> {
-        match socket {
-            service_binding::Listener::Unix(listener) => {
-                let listener = UnixListener::from_std(listener, &Default::default())?;
-                Ok(tokio::run(handle_clients!(self, listener)))
-            }
-            service_binding::Listener::Tcp(listener) => {
-                let listener = TcpListener::from_std(listener, &Default::default())?;
-                Ok(tokio::run(handle_clients!(self, listener)))
+#[async_trait]
+pub trait Agent: 'static + Sync + Send + Sized {
+    type Error: fmt::Debug + Send + Sync;
+
+    async fn handle(&self, message: Message) -> Result<Message, Self::Error>;
+
+    async fn listen<S>(self, socket: S) -> Result<(), Box<dyn Error + Send + Sync>>
+    where
+        S: ListeningSocket + fmt::Debug + Send,
+    {
+        info!("Listening; socket = {:?}", socket);
+        let arc_self = Arc::new(self);
+
+        loop {
+            match socket.accept().await {
+                Ok(socket) => {
+                    let agent = arc_self.clone();
+                    let mut session = Session::new(agent, socket);
+
+                    tokio::spawn(async move {
+                        let _ = session.handle_socket().await;
```

Comment:
I think we should `if let Err(e) ...` and print some error (with `error!`) as this silent consumption of errors could bite us later... :thinking:

Reply:
huum, if there is an error, this is a protocol error, and there isn't much to do except drop the connection. I can log the error though.

Reply:
Yep, we're on top of the "stack" here so nothing to do to recover but actually logging the potential error is exactly what I had in mind (so that we're not missing anything).

Reply:
I'll fix that in a new PR.